core/state/reader.go

```diff
@@ -48,6 +48,17 @@ type Reader interface {
 	// - The returned storage slot is safe to modify after the call
 	Storage(addr common.Address, slot common.Hash) (common.Hash, error)
 
+	// ContractCode returns the code associated with a particular account.
+	//
+	// - It returns an error to indicate code doesn't exist
+	// - The returned code is safe to modify after the call
```

Comment:
I want the docs to be _really_ clear on semantics here. Here, you say that an error is returned if `code doesn't exist`. So, does it return errors if you ask for the contract code of anything which has no code? Or only if the account does not exist in the trie? 

I mean, there are more points that could be clarified too: 
How about asking for the code of a contract which is currently marked as selfdestructed? How about during initcode execution, or just after creation? 

Also, the implementations  (snapshot) seem to return error just because they are unable to provide the code. That seems like a spec-flaw. The API doesn't allow the implementor to signal "I'm unable to do this", thus forcing it to signal "the code doesn't exist", which is not necessarily the case. An API-wart, IMO

Reply:
> I want the docs to be really clear on semantics here. Here, you say that an error is returned if code doesn't exist. So, does it return errors if you ask for the contract code of anything which has no code? Or only if the account does not exist in the trie?

I'm not sure how to formulate this better. What happens is we query disk by code hash. Address is not used at all. I believe it is kept in the (cachingDB) interface because of verkle.

So the semantic is return value only if code with that hash is found and it has size > 0.

Reply:
Then we should not have `address` in there... ?

Reply:
> Also, the implementations (snapshot) seem to return error just because they are unable to provide the code. That seems like a spec-flaw. The API doesn't allow the implementor to signal "I'm unable to do this", thus forcing it to signal "the code doesn't exist", which is not necessarily the case. An API-wart, IMO

Right. I have now added a method "SupportsCodeQuery() bool" to the interface which should return true if a reader allows you to fetch code. I'm not particularly fond of the name

Reply:
> Then we should not have address in there... ?

Summoning @rjl493456442 as I tried to stay consistent with state.Database

Reply:
The address was added for verkle. In verkle, the contract code is tight with the contract.